client/DESIGN.md

```diff
@@ -0,0 +1,292 @@
+
+Here is the method for determining the most restrictive parameter set:
+
+* **read\_timestamp**: The most restritive read\_timestamp is the minimum of the given read\_timestamp values (i.e., min(cache, transaction) read\_timestamps)
+* **max\_age**: The most restrictive max\_age is the minimum of all given max\_age values
+* **no\_cache**: The most restrictive no\_cache value is the disjunction (boolean "or") of all given no\_cache values.  (If no\_cache == True, the cache must verify with the server by checking the value timestamp it has the most up-to-date value.)
+
+### write
+    write(
+        table: string, 
+        op_List: list): value (JSON) throws StaleException  
+
+The write method sends the op\_List to the DB server as a batch write.  
+
+If the server accepts the batch write, the write method also updates the cache to reflect the newly written values.
+
+* **Question (2)**: What is the expected format of the op\_List?  (What are the possible types of its elements?)
+
+* **Question (3)**: Should the write method update the cache upon a successful write (with the newly written values)?
+
+Otherwise, the write method throws a stale exception, including the value timestamp in the DB that caused the write reject if it is available.
+
+## HTTP Message Formatting
+First, we specify the default and custom headers our HTTP messages utilize.  Then we describe the expected forms of requests/responses between the cache and TreodeDB server.
+
+### Use of HTTP Headers
+
+In our use of HTTP headers, the term *age* refers to time since the user-specified read\_timestamp.  Specifically, every TreodeDB entry has both a *value timestamp (V_t)* and *read timestamp (R_t)*.  We maintain as an invariant that the given key contained the entry's value between [V\_t, R\_t].
+
+We use the following HTTP headers:
+
+* **max-age**: The max number of seconds between our given read\_timestamp and the read\_timestamp of the entry, i.e., (input read\_timestamp) - (entry read\_timestamp) <= max-age.  In other words, we are willing to read a value that could be stale by at most max-age seconds. 
+* **no-cache**: Forces cache to send a request to the DB server to retrieve the requested value.  The result cannot just be returned from cache.
+* **Read-TxClock**: Contains the user's specified read\_timestamp.  The value returned by the DB server will be the most recent value older than the specified read\_timestamp value.  Note that Read-TxClock is an integer; it is specified in milliseconds to clarify that it should **not** be manipulated as a date.
```

Comment:
It's a 64-bit value. Let's call it long, and occasionally let's call it long (64 bits). It is called long above, but here its called integer. Let's be very consistent, and occasionally add the additional info, to avoid confusion.  Also, it's microseconds.

> should **not** be manipulated as a date

I think we should clearly explain when the application may use it as a date, and when it must not treat it as a date. The application may display it in user readable form for the user only. The application must always manipulate it internally as an 64-bit value, so that it never risks rounding to seconds.